integration-tests/itest/deploy_test.go

```diff
@@ -0,0 +1,306 @@
+package main_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"regexp"
+	"strings"
+	"time"
+
+	"os/exec"
+
+	"lightbend.com/cloudflow/itest/cli"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// Assumptions:
+// In this test, we are going to deploy the test application to an existing cluster.
+// For this first step, we assume that the cluster is correctly deployed and that kubectl is configured to talk to it.
+// The application is already built
+// We also assume that the test is going to be executed manually.
+// Automation will be the next step
+
+const (
+	JsonTokenSrc    = "/keybase/team/assassins/gcloud/pipelines-serviceaccount-key-container-registry-read-write.json"
+	ShortTimeout    = 60  // seconds
+	LongTimeout     = 240 // seconds
+	InitialWaitTime = "30s"
+)
+
+var swissKnifeApp = cli.App{
+	Image: "eu.gcr.io/bubbly-observer-178213/swiss-knife:189-277e424",
+	Name:  "swiss-knife",
+}
+
+var _ = Describe("Application deployment", func() {
+	Context("the cluster is clean for testing", func() {
+		It("should not have the test app", func() {
+			err := ensureAppNotDeployed(swissKnifeApp)
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
+
+	Context("when I deploy an application", func() {
+		It("should start a deployment", func() {
+			jsonToken := getToken()
+			output, err := cli.Deploy(swissKnifeApp, jsonToken)
+			Expect(err).NotTo(HaveOccurred())
+			expected := "Deployment of application `" + swissKnifeApp.Name + "` has started."
+			Expect(output).To(ContainSubstring(expected))
+			waitTime, _ := time.ParseDuration(InitialWaitTime)
+			time.Sleep(waitTime) // this wait is needed to let the application deploy
+		})
+
+		It("should be in the list of applications in the cluster", func() {
+			list, err := cli.ListApps()
+			Expect(err).NotTo(HaveOccurred())
+			appNames := listAppNames(list)
+			Expect(appNames).To(ContainElement(swissKnifeApp.Name))
+		})
+
+		It("should get to a 'running' status, eventually", func(done Done) {
+			// TODO: check status flag once it's fixed
+			status, err := checkStatusIs(swissKnifeApp, "Running")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(status).To(Equal("Running"))
+			close(done)
+		}, LongTimeout)
+	})
+
+	Context("A deployed test application that uses akka, spark, and flink", func() {
+		It("should contain a spark process", func() {
+			status, err := cli.Status(swissKnifeApp)
+			Expect(err).NotTo(HaveOccurred())
+			streamlets := cli.GetStreamlets(&status)
+			Expect(streamlets).To(ContainElement("spark-process"))
```

Comment:
would it be a good idea to make the test less brittle by checking for the runtime instead of the hardcoded streamlet name?

Reply:
How could we do that with `kubectl cloudflow` ? 
Maybe the answer is that we should have a column "backend" in the status report. 
This is how it looks like right now:
```
$ kubectl cloudflow status swiss-knife
Name:             swiss-knife
Namespace:        swiss-knife
Version:          189-277e424
Created:          2020-03-24 12:53:12 +0100 CET
Status:           Pending

STREAMLET         POD                              STATUS            RESTARTS          READY             
spark-process     swiss-knife-spark-process-driver Unknown           0                 
ingress           swiss-knife-ingress-driver       Unknown           0   
```

Reply:
I agree part of our end to end testing is about testing the user interaction. IMO that shouldn't mean we cannot use the lower level aspects of kubernetes, all accessible through kubectl, to make our tests more powerful.

that said, re-reading your test I agree that this usage of hardcoded streamlet names is fairly acceptable. The biggest problem for me is the high decoupling between the test application and the tests, which makes it likely that changes in the test application will frequently break these tests. I was looking for a way to make that less likely for innocent changes (as in: they would not have a breaking effect in the real world) such as changing the name of a streamlet.

Reply:
I changed the test structure to make clear that this test is checking the `status` output and not the runtime. It would be good to have a "spark" and "flink" suites that check specific characteristics and behavior for each backend.